File: core/src/main/java/org/hisp/dhis/client/sdk/core/trackedentity/TrackedEntityInstanceService.java
/*
 * Copyright (c) 2016, University of Oslo
 *
 * All rights reserved.
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * Neither the name of the HISP project nor the names of its contributors may
 * be used to endorse or promote products derived from this software without
 * specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package org.hisp.dhis.client.sdk.core.trackedentity;

import org.hisp.dhis.client.sdk.core.common.StateStore;
import org.hisp.dhis.client.sdk.core.common.utils.CodeGenerator;
import org.hisp.dhis.client.sdk.core.relationship.RelationshipStore;
import org.hisp.dhis.client.sdk.models.common.state.Action;
import org.hisp.dhis.client.sdk.models.organisationunit.OrganisationUnit;
import org.hisp.dhis.client.sdk.models.relationship.Relationship;
import org.hisp.dhis.client.sdk.models.relationship.RelationshipType;
import org.hisp.dhis.client.sdk.models.trackedentity.TrackedEntity;
import org.hisp.dhis.client.sdk.models.trackedentity.TrackedEntityInstance;
import org.hisp.dhis.client.sdk.models.utils.Preconditions;

import java.util.List;

public final class TrackedEntityInstanceService implements ITrackedEntityInstanceService {

    private final TrackedEntityInstanceStore trackedEntityInstanceStore;
    private final RelationshipStore relationshipStore;
    private final StateStore stateStore;

    public TrackedEntityInstanceService(TrackedEntityInstanceStore trackedEntityInstanceStore,
                                        RelationshipStore relationshipStore, StateStore
                                                stateStore) {
        this.trackedEntityInstanceStore = trackedEntityInstanceStore;
        this.relationshipStore = relationshipStore;
        this.stateStore = stateStore;
    }

    @Override
    public TrackedEntityInstance get(String uid) {
        TrackedEntityInstance trackedEntityInstance = trackedEntityInstanceStore.queryByUid(uid);
        Action action = stateStore.queryActionForModel(trackedEntityInstance);

        if (!Action.TO_DELETE.equals(action)) {
            return trackedEntityInstance;
        }

        return null;
    }

    @Override
    public TrackedEntityInstance create(TrackedEntity trackedEntity, OrganisationUnit
            organisationUnit) {
        Preconditions.isNull(trackedEntity, "Tracked entity must not be null");
        Preconditions.isNull(organisationUnit, "Organisation unit must not be null");

        TrackedEntityInstance trackedEntityInstance = new TrackedEntityInstance();
        trackedEntityInstance.setOrgUnit(organisationUnit.getUId());
        trackedEntityInstance.setTrackedEntity(trackedEntity.getUId());
        trackedEntityInstance.setTrackedEntityInstanceUid(CodeGenerator.generateCode());
        save(trackedEntityInstance);

        return trackedEntityInstance;
    }

    @Override
    public boolean addRelationship(TrackedEntityInstance trackedEntityInstanceA,
                                   TrackedEntityInstance trackedEntityInstanceB, RelationshipType
                                               relationshipType) {
        Preconditions.isNull(trackedEntityInstanceA, "Tracked entity instance A must not be null");
        Preconditions.isNull(trackedEntityInstanceB, "Tracked entity instance B must not be null");
        Preconditions.isNull(relationshipType, "Relationship type must not be null");

        List<Relationship> existingRelationships = trackedEntityInstanceA.getRelationships();
        for (Relationship existingRelationship : existingRelationships) {
            if (existingRelationship.getTrackedEntityInstanceB().getTrackedEntityInstanceUid()
                    .equals(trackedEntityInstanceB.getTrackedEntityInstanceUid()) &&
                    relationshipType.getUId().equals(existingRelationship.getRelationship())) {
                return false;
            }
        }
        Relationship relationship = new Relationship();
        relationship.setTrackedEntityInstanceA(trackedEntityInstanceA);
        relationship.setTrackedEntityInstanceB(trackedEntityInstanceB);
        relationship.setRelationship(relationship.getRelationship());
        relationshipStore.insert(relationship);
        trackedEntityInstanceA.getRelationships().add(relationship);
        trackedEntityInstanceB.getRelationships().add(relationship);
        save(trackedEntityInstanceA);
        save(trackedEntityInstanceB);
        return true;
    }

    @Override
    public boolean removeRelationship(Relationship relationship) {
        Preconditions.isNull(relationship, "Relationship must not be null");
        relationshipStore.delete(relationship);
        relationship.getTrackedEntityInstanceA().getRelationships().remove(relationship);
        relationship.getTrackedEntityInstanceB().getRelationships().remove(relationship);
        save(relationship.getTrackedEntityInstanceA());
        save(relationship.getTrackedEntityInstanceB());
        return true;
    }

    @Override
    public TrackedEntityInstance get(long id) {
        TrackedEntityInstance trackedEntityInstance = trackedEntityInstanceStore.queryById(id);
        Action action = stateStore.queryActionForModel(trackedEntityInstance);

        if (!Action.TO_DELETE.equals(action)) {
            return trackedEntityInstance;
        }

        return null;
    }

    @Override
    public List<TrackedEntityInstance> list() {
        return stateStore.queryModelsWithActions(TrackedEntityInstance.class, Action.SYNCED,
                Action.TO_UPDATE, Action.TO_POST);
    }

    @Override
    public boolean remove(TrackedEntityInstance object) {
        Preconditions.isNull(object, "trackedEntityInstance argument must not be null");
        if (!trackedEntityInstanceStore.delete(object)) {
            return false;
        }
        return stateStore.deleteActionForModel(object);
    }

    @Override
    public boolean save(TrackedEntityInstance object) {
        Preconditions.isNull(object, "Tracked entity instance must not be null");

        if (!trackedEntityInstanceStore.save(object)) {
            return false;
        }

        // TODO check if object was created earlier (then set correct flag)
        Action action = stateStore.queryActionForModel(object);

        if (action == null || Action.TO_POST.equals(action)) {
            return stateStore.saveActionForModel(object, Action.TO_POST);
        } else {
            return stateStore.saveActionForModel(object, Action.TO_UPDATE);
        }

    }
}

File: core/src/main/java/org/hisp/dhis/android/network/trackedentityinstance/TrackedEntityInstanceService.kt
/*
 *  Copyright (c) 2004-2025, University of Oslo
 *  All rights reserved.
 *
 *  Redistribution and use in source and binary forms, with or without
 *  modification, are permitted provided that the following conditions are met:
 *  Redistributions of source code must retain the above copyright notice, this
 *  list of conditions and the following disclaimer.
 *
 *  Redistributions in binary form must reproduce the above copyright notice,
 *  this list of conditions and the following disclaimer in the documentation
 *  and/or other materials provided with the distribution.
 *  Neither the name of the HISP project nor the names of its contributors may
 *  be used to endorse or promote products derived from this software without
 *  specific prior written permission.
 *
 *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 *  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 *  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
 *  ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 *  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
 *  ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 *  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 *  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package org.hisp.dhis.android.network.trackedentityinstance

import org.hisp.dhis.android.core.trackedentity.TrackedEntityInstance
import org.hisp.dhis.android.core.arch.api.HttpServiceClient
import org.hisp.dhis.android.network.common.fields.Fields

@Suppress("LongParameterList")
internal class TrackedEntityInstanceService(private val client: HttpServiceClient) {
    suspend fun postTrackedEntityInstances(
        trackedEntityInstances: TrackedEntityInstancePayload?,
        strategy: String?,
    ): TEIWebResponseDTO {
        return client.post {
            url(TRACKED_ENTITY_INSTANCES)
            parameters {
                attribute(STRATEGY, strategy)
            }
            body(trackedEntityInstances)
        }
    }

    suspend fun getTrackedEntityInstance(
        trackedEntityInstance: String,
        orgUnitMode: String?,
        fields: Fields<TrackedEntityInstance>,
        includeAllAttributes: Boolean,
        includeDeleted: Boolean,
    ): TrackedEntityInstancePayload {
        return client.get {
            url(TRACKED_ENTITY_INSTANCES)
            parameters {
                fields(fields)
                attribute(TRACKED_ENTITY_INSTACE, trackedEntityInstance)
                attribute(OU_MODE, orgUnitMode)
                attribute(INCLUDE_ALL_ATTRIBUTES, includeAllAttributes)
                attribute(INCLUDE_DELETED, includeDeleted)
            }
        }
    }

    suspend fun getSingleTrackedEntityInstance(
        trackedEntityInstanceUid: String,
        orgUnitMode: String?,
        program: String?,
        programStatus: String?,
        programStartDate: String?,
        fields: Fields<TrackedEntityInstance>,
        includeAllAttributes: Boolean,
        includeDeleted: Boolean,
    ): TrackedEntityInstanceDTO {
        return client.get {
            url("$TRACKED_ENTITY_INSTANCES/$trackedEntityInstanceUid")
            parameters {
                fields(fields)
                attribute(OU_MODE, orgUnitMode)
                attribute(PROGRAM, program)
                attribute(PROGRAM_STATUS, programStatus)
                attribute(PROGRAM_START_DATE, programStartDate)
                attribute(INCLUDE_ALL_ATTRIBUTES, includeAllAttributes)
                attribute(INCLUDE_DELETED, includeDeleted)
            }
        }
    }

    suspend fun getTrackedEntityInstances(
        trackedEntityInstances: String?,
        orgUnits: String?,
        orgUnitMode: String?,
        program: String?,
        programStatus: String?,
        programStartDate: String?,
        fields: Fields<TrackedEntityInstance>,
        order: String?,
        paging: Boolean,
        page: Int?,
        pageSize: Int?,
        lastUpdatedStartDate: String?,
        includeAllAttributes: Boolean,
        includeDeleted: Boolean,
    ): TrackedEntityInstancePayload {
        return client.get {
            url(TRACKED_ENTITY_INSTANCES)
            parameters {
                fields(fields)
                attribute(TRACKED_ENTITY_INSTACE, trackedEntityInstances)
                attribute(OU, orgUnits)
                attribute(OU_MODE, orgUnitMode)
                attribute(PROGRAM, program)
                attribute(PROGRAM_STATUS, programStatus)
                attribute(PROGRAM_START_DATE, programStartDate)
                attribute(ORDER, order)
                attribute(LAST_UPDATED_START_DATE, lastUpdatedStartDate)
                attribute(INCLUDE_ALL_ATTRIBUTES, includeAllAttributes)
                attribute(INCLUDE_DELETED, includeDeleted)
                paging(paging)
                page(page)
                pageSize(pageSize)
            }
        }
    }

    suspend fun query(
        trackedEntityInstance: String?,
        orgUnit: String?,
        orgUnitMode: String?,
        program: String?,
        programStage: String?,
        programStartDate: String?,
        programEndDate: String?,
        enrollmentStatus: String?,
        programIncidentStartDate: String?,
        programIncidentEndDate: String?,
        followUp: Boolean?,
        eventStartDate: String?,
        eventEndDate: String?,
        eventStatus: String?,
        trackedEntityType: String?,
        filter: List<String?>?,
        assignedUserMode: String?,
        lastUpdatedStartDate: String?,
        lastUpdatedEndDate: String?,
        order: String?,
        paging: Boolean,
        page: Int?,
        pageSize: Int?,
    ): SearchGridDTO {
        return client.get {
            url("$TRACKED_ENTITY_INSTANCES/query")
            parameters {
                filter(filter)
                attribute(TRACKED_ENTITY_INSTACE, trackedEntityInstance)
                attribute(OU, orgUnit)
                attribute(OU_MODE, orgUnitMode)
                attribute(PROGRAM, program)
                attribute(PROGRAM_STAGE, programStage)
                attribute(PROGRAM_START_DATE, programStartDate)
                attribute(PROGRAM_END_DATE, programEndDate)
                attribute(PROGRAM_STATUS, enrollmentStatus)
                attribute(PROGRAM_INCIDENT_START_DATE, programIncidentStartDate)
                attribute(PROGRAM_INCIDENT_END_DATE, programIncidentEndDate)
                attribute(FOLLOW_UP, followUp)
                attribute(EVENT_START_DATE, eventStartDate)
                attribute(EVENT_END_DATE, eventEndDate)
                attribute(EVENT_STATUS, eventStatus)
                attribute(TRACKED_ENTITY_TYPE, trackedEntityType)
                attribute(ASSIGNED_USER_MODE, assignedUserMode)
                attribute(LAST_UPDATED_START_DATE, lastUpdatedStartDate)
                attribute(LAST_UPDATED_END_DATE, lastUpdatedEndDate)
                attribute(ORDER, order)
                paging(paging)
                page(page)
                pageSize(pageSize)
            }
        }
    }

    companion object {
        const val TRACKED_ENTITY_INSTANCES = "trackedEntityInstances"
        const val TRACKED_ENTITY_INSTACE = "trackedEntityInstance"
        const val OU = "ou"
        const val OU_MODE = "ouMode"
        const val PROGRAM = "program"
        const val PROGRAM_STAGE = "programStage"
        const val PROGRAM_START_DATE = "programStartDate"
        const val PROGRAM_END_DATE = "programEndDate"
        const val PROGRAM_STATUS = "programStatus"
        const val PROGRAM_INCIDENT_START_DATE = "programIncidentStartDate"
        const val PROGRAM_INCIDENT_END_DATE = "programIncidentEndDate"
        const val FOLLOW_UP = "followUp"
        const val EVENT_STATUS = "eventStatus"
        const val EVENT_START_DATE = "eventStartDate"
        const val EVENT_END_DATE = "eventEndDate"
        const val TRACKED_ENTITY_TYPE = "trackedEntityType"
        const val INCLUDE_ALL_ATTRIBUTES = "includeAllAttributes"
        const val FILTER = "filter"
        const val STRATEGY = "strategy"
        const val LAST_UPDATED_START_DATE = "lastUpdatedStartDate"
        const val LAST_UPDATED_END_DATE = "lastUpdatedEndDate"
        const val INCLUDE_DELETED = "includeDeleted"
        const val ASSIGNED_USER_MODE = "assignedUserMode"
        const val ORDER = "order"
        const val DEFAULT_PAGE_SIZE = 20
    }
}
